feat: Add dot notation support for all JWT fields

colesmcintosh opened the pull request:
## Title

feat: Add dot notation support for all JWT fields

## Relevant issues

Addresses the need for nested JWT claim access across all JWT fields in LiteLLM proxy authentication.

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## Type

🆕 New Feature

## Changes

### Summary
Extended JWT authentication to support dot notation for accessing nested claims across all JWT fields, not just `roles_jwt_field`. This enables organizations with complex JWT structures to map nested claims to LiteLLM user/team/org identifiers.

### Implementation Details

**Core Changes:**
- **Updated `handle_jwt.py`**: Modified all JWT field access methods to use `get_nested_value` instead of direct dictionary indexing
- **Enhanced `get_team_id`**: Fixed fallback logic for `team_id_default` to work correctly with nested field access
- **Maintained backward compatibility**: All existing flat field configurations continue to work unchanged

**Supported Fields with Dot Notation:**
- `team_id_jwt_field` (e.g., `"tenant.team_id"`)
- `team_ids_jwt_field` (e.g., `"groups"`, `"teams.ids"`)
- `user_id_jwt_field` (e.g., `"user.sub"`)
- `user_email_jwt_field` (e.g., `"user.email"`)
- `org_id_jwt_field` (e.g., `"organization.id"`)
- `object_id_jwt_field` (e.g., `"profile.object_id"`)
- `end_user_id_jwt_field` (e.g., `"customer.end_user_id"`)

### Testing
Added comprehensive test coverage in `tests/proxy_unit_tests/test_jwt.py`:

1. **`test_nested_jwt_field_access`**: Tests all JWT fields with both nested and flat access patterns
2. **`test_nested_jwt_field_missing_paths`**: Verifies proper fallback behavior when nested paths don't exist  
3. **`test_metadata_prefix_handling_in_nested_fields`**: Tests metadata prefix removal functionality

**Test Results:**
<img width="798" height="107" alt="Screenshot 2025-07-26 at 12 32 11 PM" src="https://github.com/user-attachments/assets/d60a8508-16b0-4a0f-881a-f8726ee85d4d" />

### Documentation
Updated `docs/my-website/docs/proxy/token_auth.md` with:
- Clear explanation that all JWT fields support dot notation
- Examples of both flat and nested JWT token structures
- Configuration examples for nested field access

### Example Usage

**Before (flat only):**
```yaml
litellm_jwtauth:
  user_id_jwt_field: "sub"
  team_id_jwt_field: "client_id"
```

**Now (supports nested):**
```yaml
litellm_jwtauth:
  user_id_jwt_field: "user.sub"
  user_email_jwt_field: "user.email"
  team_id_jwt_field: "tenant.team_id"
  org_id_jwt_field: "organization.id"
```

### Backward Compatibility
✅ All existing configurations continue to work without changes
✅ No breaking changes to existing APIs
✅ Existing tests continue to pass

colesmcintosh committed:
feat: Add dot notation support for all JWT fields

- Updated all JWT field access methods to use get_nested_value for dot notation support
- Enhanced get_team_id to properly handle team_id_default fallback with nested fields
- Added comprehensive unit tests for nested JWT field access and edge cases
- Updated documentation to reflect dot notation support across all JWT fields
- Maintains full backward compatibility with existing flat field configurations

Supported fields with dot notation:
- team_id_jwt_field, team_ids_jwt_field, user_id_jwt_field
- user_email_jwt_field, org_id_jwt_field, object_id_jwt_field
- end_user_id_jwt_field (roles_jwt_field was already supported)

Example: user_id_jwt_field: 'user.sub' accesses token['user']['sub']

colesmcintosh committed:
fix: Add type annotations to resolve mypy errors

- Add explicit type annotation for team_ids variable in get_team_ids_from_jwt
- Add type ignore comment for sentinel object return in get_team_id
- Resolves mypy errors while maintaining functionality

colesmcintosh committed:
fix: Resolve mypy type error in get_team_ids_from_jwt

- Remove explicit List[str] type annotation that conflicts with get_nested_value return type
- Simplify return logic to use 'team_ids or []' ensuring always returns List[str]
- Fixes: Incompatible types in assignment (expression has type 'list[str] | None', variable has type 'list[str]')

colesmcintosh committed:
fix: Add proper type annotation for team_ids variable

- Use Optional[List[str]] type annotation to satisfy mypy requirements
- Resolves: Need type annotation for 'team_ids' [var-annotated]
- Maintains functionality while ensuring type safety

colesmcintosh commented:
bugbot run

colesmcintosh commented:
bugbot run

krrishdholakia commented on tests/proxy_unit_tests/test_jwt.py:
move to `test_litellm/` so we can have this run on the github action

colesmcintosh committed:
Merge branch 'BerriAI:main' into feat/jwt-dot-notation-support

colesmcintosh committed:
refactor: remove outdated JWT unit tests and consolidate JWT-related functionality

- Deleted the test_jwt.py file as it contained outdated and redundant tests.
- Consolidated JWT-related tests into test_handle_jwt.py for better organization and maintainability.
- Updated tests to ensure proper functionality of JWT handling, including token validation and role mapping.
- Enhanced test coverage for JWT field access and nested claims handling.

colesmcintosh committed:
test: add comprehensive unit tests for JWT authentication

- Introduced a new test file `test_jwt.py` containing unit tests for JWT authentication.
- Implemented tests for loading configuration with custom role names, validating tokens, and handling team tokens.
- Enhanced coverage for JWT field access, nested claims, and role-based access control.
- Added fixtures for Prisma client and public JWT key generation to support testing.
- Ensured proper handling of valid and invalid tokens, including user and team scenarios.

colesmcintosh committed:
revert test_handle_jwt.py

colesmcintosh committed:
rename file

colesmcintosh committed:
test: remove outdated JWT nesting tests and add new nested field access tests

- Deleted the `test_jwt_nesting.py` file as it contained outdated tests.
- Introduced new tests in `test_handle_jwt.py` to verify nested JWT field access.
- Enhanced coverage for accessing nested values using dot notation and ensured backward compatibility with flat field names.
- Added tests for handling missing nested paths and appropriate default values.
- Improved handling of metadata prefixes in nested field access.

colesmcintosh committed:
restore file

ishaan-jaff reviewed (approved):
lgtm

colesmcintosh committed:
Merge branch 'BerriAI:main' into feat/jwt-dot-notation-support
